### ISSUE TITLE
Would you accept a custom error encoder representation?

### ISSUE THREAD

**Issue by codyaray** [getkin/kin-openapi#194]
There's a non-trivial amount of effort to transform the returned errors into something we can encode and send to the user. I'm writing a custom encoder right now for our own purposes, but wanted to know if this is something you would consider accepting upstream to share with other users in the community?

Since the exact encoding is application dependent, we write it into an intermediate "error object" that's easier to just encode into whatever media type and structure the app uses. This is an easier jumping off point for other developers.

Below I've included samples of the tests from my work. This intermediate error object can be seen in the `wantErrResponse` below. They also include some actual TODOs in my code where I think we could improve the upstream to return better errors generally.

```
//
// Basics
//

{
	name: "error - unknown path",
	args: validationArgs{
		r: badPath,
	},
	wantErrReason:   "Path was not found",
	wantErrResponse: &corev1.Error{Code: http.StatusNotFound, Message: "Path was not found"},
},

{
	name: "error - unknown method",
	args: validationArgs{
		r: badMethod,
	},
	wantErrReason:   "Path doesn't support the HTTP method",
	// TODO: By HTTP spec, this should have an Allow header with what is allowed
	// but kin-openapi doesn't provide us the requested method or path, so impossible to provide details
	wantErrResponse: &corev1.Error{Code: http.StatusMethodNotAllowed, Message: "Path doesn't support the HTTP method"},
},

<snip>

//
// Content-Type
//

{
	name: "error - unknown content-type on POST",
	args: validationArgs{
		r: unknownContentType,
	},
	wantErrReason:      "failed to decode request body",
	wantErrParseKind:   openapi3filter.KindUnsupportedFormat,
	wantErrParseReason: "unsupported content type \"application/xml\"",
	wantErrResponse:    &corev1.Error{Code: http.StatusUnsupportedMediaType, Message: "unsupported content type \"application/xml\""},
},
{
	name: "error - unsupported content-type on POST",
	args: validationArgs{
		r: unsupportedContentType,
	},
	wantErrReason:   "header 'Content-Type' has unexpected value: \"text/plain\"",
	wantErrResponse: &corev1.Error{Code: http.StatusUnsupportedMediaType, Message: "unsupported content type \"text/plain\""},
},

<snip>

//
// Query strings
//

{
	name: "error - invalid enum value for query string parameter",
	args: validationArgs{
		r: newPetstoreRequest(t, http.MethodGet, "/pet/findByStatus?status=sold&status=watdis", nil),
	},
	wantErrParam:        "status",
	wantErrParamIn:      "query",
	wantErrSchemaReason: "JSON value is not one of the allowed values",
	wantErrSchemaPath:   "/1",
	wantErrSchemaValue:  "watdis",
	wantErrResponse:     &corev1.Error{Code: http.StatusBadRequest,
		Message: "JSON value is not one of the allowed values",
		Details: []string{"Value 'watdis' at /1 must be one of: available, pending, sold"},
		Source: &corev1.ErrorSource{Parameter: "status"}},
},

<snip>

//
// Request bodies
//

{
	name: "error - missing required object attribute",
	args: validationArgs{
		r: newPetstoreRequest(t, http.MethodPost, "/pet", bytes.NewBufferString(`{"name":"Bahama"}`)),
	},
	wantErrReason:       "doesn't match the schema",
	wantErrSchemaReason: "Property 'photoUrls' is missing",
	wantErrSchemaValue:  map[string]string{"name": "Bahama"},
	// TODO update kin-openapi errors to include the 'reversePath' here so we can add a source
	// We'd like this to include Source: &corev1.ErrorSource{Pointer: "/photoUrls"}},
	wantErrResponse: &corev1.Error{Code: http.StatusUnprocessableEntity, Message: "Property 'photoUrls' is missing"},
	// TODO: test this scenario with nested structures. I think the message will give the attribute but not path
},

<snip>

//
// Path params
//

{
	name: "error - wrong path param type",
	args: validationArgs{
		r: newPetstoreRequest(t, http.MethodGet, "/pet/NotAnInt", nil),
	},
	wantErrParam:       "petId",
	wantErrParamIn:     "path",
	wantErrParseKind:   openapi3filter.KindInvalidFormat,
	wantErrParseValue:  "NotAnInt",
	wantErrParseReason: "an invalid integer",
	wantErrResponse:    &corev1.Error{Code: http.StatusNotFound, Message: "Resource not found with 'petId' value: NotAnInt"},
},
```

This also shows that the intermediate encoding is opinionated, so worst case a developer can fork this encoder as a starting point, rather than working from scratch.

**Comment by fenollp** [comment]
I am definitely for a more precise error types. Not necessarily a generic error type with string fields but we can start this way which is a good middle ground (fields can be added without breaking API contracts): https://github.com/xeipuuv/gojsonschema/blob/b537c054d4b486d62f64a76b8d48fdf38a0ea99d/errors.go

So let's aim for very precise error types wrapping enough context such that by using the error and the failed function call input values one can build a custom error handler using a type switch (as shown in link) or `errors.Is` / `errors.As`.

I will definitely support you in this endeavour!

**Comment by codyaray** [comment]
I was actually trying to move from relatively precise error types (and nested errors) into a single flat generic error type suitable for encoding. Right now, I suspect (but you would know better) that most users simplify call `Error()` and use the colon-delimited string representation as their error message.

Basically, I want to find a middle ground between the current "relatively precise error types" and this basic `Error()` message. Something with granular fields for structured error encoding, but easier to use than type switching on a ton of specific errors. Specifically, I'm personally looking for something that can map into a structure like https://jsonapi.org/format/#error-objects (but obviously not requiring this specific format or anything).

**Comment by fenollp** [comment]
I fail to see your point here:
* Fine-grained error types still provide `Error() string` and
* some generic Stringer can be implemented by having all these fine grained error types implement the formatter of your choice (some interface).
No huge and repeated type switches here, no?

**Comment by codyaray** [comment]
@fenollp I went ahead and opened a PR #197 to try making this a bit more concrete. 

But I'd also like to just make my use case and base assumptions a bit clearer here. These changes are primarily intended for API service providers (like myself) who want to validate all incoming API requests and outgoing API responses against the OpenAPI spec.

As an API service provider, I assume that
* your API accepts requests in one or more media types according to the `Content-Type` header. Example media types include `application/json`, `application/xml`, `application/vnd.api+json`
* your API responds in one or more media types according to the `Accept` header.
* your API already has some standard error object which can be represented in any of the above media types. For example, your error object may be represented in JSONAPI like:
```
{"errors":[{
    "id": "06120fc4-660d-4fd9-b6b0-42e7a5c96c5b",
    "status": 422,
    "code": "parameter_invalid_empty",
    "title": "Request body is missing a property",
    "detail": "Property 'name' is missing",
    "source": {"pointer": "/category/tags/0/name"}
}]}
```

I hope this makes my intent clearer: I want to easily translate the kin-openapi errors into my error object, which already has encoders for the various media types my API supports.

If this is still a bit too high level, take a look at how I'm actually having to parse the kin-openapi errors right now:
* [checking type assertions](https://github.com/getkin/kin-openapi/pull/197/files#diff-ee2d579818e5124734fb3ce2cd7e87acR38)
* [switching on `Reason` strings](https://github.com/getkin/kin-openapi/pull/197/files#diff-ee2d579818e5124734fb3ce2cd7e87acR54)
* [regex parsing the Reason string to get at bits of data ](https://github.com/getkin/kin-openapi/pull/197/files#diff-ee2d579818e5124734fb3ce2cd7e87acR158)
* [the JSONPointer is super helpful, but I have to "complete" it](https://github.com/getkin/kin-openapi/pull/197/files#diff-ee2d579818e5124734fb3ce2cd7e87acR155)

There's definitely some improvements we can do "inside" the core error generation to remove the need for some of this post-processing, but I wasn't sure how "deep" you wanted to push changes for this particular use case (especially if they're backward incompatible, like changing the `JSONPointer()`).

**Comment by codyaray** [comment]
Addressed in #197
